src/Module.zig

```diff
@@ -1848,6 +1848,15 @@ pub const File = struct {
         if (amt != stat.size)
             return error.UnexpectedEndOfFile;
 
+        const resolved_path = try file.fullPath(gpa);
+        defer gpa.free(resolved_path);
+        const real_path = try std.fs.realpathAlloc(gpa, resolved_path);
+        if (!mem.endsWith(u8, real_path, file.sub_file_path)) {
+            file.source = @ptrCast([:0]const u8, real_path);
+            return error.ImportPathMismatch;
+        }
+        defer gpa.free(real_path);
```

Comment:
Hi @Vexu, thank you for clarifying your point here.

My reasoning was as follows:

1. Unlike `embedFile` the `importFile` method does not actually load the file immediately. This is deferred to the thread that handles the parsing of the zig file. Therefore I assume we should do the least amount of (IO) work on the importing thread.
2. We want to load the file first before checking for the path naming issue. See my comment on `embedFile` below.

Both reasons lead me to implement the check only when the `File` is loaded. Please correct me if my assumptions are wrong here.

Reply:
>1. Unlike `embedFile` the `importFile` method does not actually load the file immediately. This is deferred to the thread that handles the parsing of the zig file. Therefore I assume we should do the least amount of (IO) work on the importing thread.

By the time we get to `Sema.importFile` all relevant files should already have been loaded by AstGen.

>2. We want to load the file first before checking for the path naming issue. See my comment on `embedFile` below.

I'm not sure why we would want to do that but it doesn't apply here anyways since file would already be loaded.

Reply:
Sorry for being stubborn, but I have a hard time to follow your thoughts.

From analysing the code it looks like this:
* In `Module` we have `fn astGenFile`. This method seems to be the only place an imported zig file is loaded, after checking for compilation cache and so on.
* This method is invoked from `fn workerAstGenFile` on `Compilation`, which delegates error handling to `fn reportRetryableAstGenError`.

From the other side:
* In `Sema` we have `fn zirImport` that calls `fn importFile` from `Module` which creates an "unloaded" File instance and then creates a `file.root_decl` that is the referenced thing that is returned. But no actual file loading is started here.

For sure we can move the `real_path` check to the `zirImport` side. But this would also trigger the check even if we have already cached the compiled binary for this file. I don't the see advantage.